Add documentation for predict API and uv as env manager (#14039)
mlflow/mlflow: docs/source/models.rst

Validate Models before Deployment
---------------------------------

After logging your model with MLflow Tracking, it is highly recommended to validate the model locally before deploying it to production.
The :py:func:`mlflow.models.predict` API provides a convenient way to test your model in a virtual environment, offering isolated execution and several advantages:

* Model dependencies validation: The API helps ensure that the dependencies logged with the model are correct and sufficient by executing the model with an input example in a virtual environment.
  For more details, refer to :ref:`Validating Environment for Prediction <validating-environment-for-prediction>`.
* Input data validation: The API can be used to validate the input data interacts with the model as expected by simulating the same data processing during model serving.
  Ensure that the input data is a valid example that aligns with the pyfunc model’s predict function requirements.

> * Extra environment variables validation: By specifying the `extra_envs` parameter, you can test whether additional environment variables are needed for the model to run successfully.

Reviewer: ```suggestion
* Extra environment variables validation: By specifying the `extra_envs` parameter, you can test whether additional environment variables are required for the model to run successfully.
```

> * Extra environment variables validation: By specifying the `extra_envs` parameter, you can test whether additional environment variables are needed for the model to run successfully.

Reviewer 2: ```suggestion
* Extra environment variables validation: By specifying the ``extra_envs`` parameter, you can test whether additional environment variables are needed for the model to run successfully.
```

nit: Single backtick becomes italic in sphinx. Same for other places.

> * Extra environment variables validation: By specifying the `extra_envs`` parameter, you can test whether additional environment variables are needed for the model to run successfully.

Reviewer 3: ```suggestion
* Extra environment variables validation: By specifying the `extra_envs` parameter, you can test whether additional environment variables are needed for the model to run successfully.
```

  Note that all existing environment variables in `os.environ` are automatically passed into the virtual environment.

.. code-block:: python

    import mlflow

    class MyModel(mlflow.pyfunc.PythonModel):
        def predict(self, context, model_input, params=None):
            return model_input

    with mlflow.start_run():
        model_info = mlflow.pyfunc.log_model(
            "model",
            python_model=MyModel(),
            input_example=["a", "b", "c"],
        )

    mlflow.models.predict(
        model_uri=model_info.model_uri,
        input_data=["a", "b", "c"],
        pip_requirements_override=["..."],
        extra_envs={"MY_ENV_VAR": "my_value"},
    )

> Different environment managers

Reviewer: ```suggestion
Environment managers
```

do we need `diffrent`?

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> The :py:func:`mlflow.models.predict` API supports different environment managers to create the virtual environment for prediction. The following environment managers are supported:

Reviewer: ```suggestion
The :py:func:`mlflow.models.predict` API supports the following environment managers to create the virtual environment for prediction:
```

* virtualenv: the default environment manager which creates a virtual environment using `virtualenv` and `pyenv`.

> * uv: uses `uv` to create a virtual environment. For more details, refer to `uv documentation <https://docs.astral.sh/uv/>`_.

Reviewer 3: is it intentional to give more information of uv than other package managers?

Author: I can add links for other package managers as well!

Reviewer 3: I think that'd be helpful unless we intentionally encourage uv!

> * uv: uses `uv` to create a virtual environment. For more details, refer to `uv documentation <https://docs.astral.sh/uv/>`_.

Reviewer: Let's mention this is experimental.

> * `virtualenv <https://virtualenv.pypa.io/en/latest/>`_: the default environment manager which creates a virtual environment using `virtualenv` and `pyenv`.
> * `uv <https://docs.astral.sh/uv/>`_: uses `uv` to create a virtual environment. **This is an experimental feature since MLflow 2.20.0.**
> * `conda <https://docs.conda.io/projects/conda/>`_: uses conda to create environment.

Reviewer 2: ```suggestion
* `virtualenv <https://virtualenv.pypa.io/en/latest/>`_: The default environment manager.
* `uv <https://docs.astral.sh/uv/>`_:  An **extremely fast** environment manager written in Rust. This is an experimental feature since MLflow 2.20.0.
* `conda <https://docs.conda.io/projects/conda/>`_
```

nit: It sounds a bit verbose to say "A is a env manager that uses A"

> * local: uses the current environment to run the model. Note that `pip_requirements_override` is not supported in this mode.

Reviewer: ```suggestion
* `local`: uses the current environment to run the model. Note that `pip_requirements_override` is not supported in this mode.
```

Author: If we add this, local would display as _uv_, but others are normal just with links

.. tip::

>     Starting from MLflow 2.20.0, the `uv` environment manager is available. It is a extremely fast and lightweight environment manager.

Reviewer: ```suggestion
    Starting from MLflow 2.20.0, the `uv` environment manager is available. It is a extremely fast.
```

    Run `pip install uv` to install uv, or refer to `uv installation guidance <https://docs.astral.sh/uv/getting-started/installation>`_ for other installation methods.

> Example of using `uv` as the environment manager:

Reviewer: ```suggestion
Example of making predictions with `uv`:
```

Author: Emm it's not "making predictions with uv", but "using uv to create virtualenv" 🤔 This sounds like we're using uv to do prediction

> Example of using `uv` to create virtual environment for prediction:

Reviewer 2: ```suggestion
Example of using `uv` to create a virtual environment for prediction:
```
